### PR TITLE
Dev: Add commandq to iiowidgets

### PR DESCRIPTION
**Pull request by Andrei-Fabian-Pop**
Add 2 new data strategies for manipulating channel and device data through iioutils/commandqueue. Also add possibility of having const data for optional data, meaning that the optional data does not have to always be read through iio.

### CODE REVIEW

File: iio-widgets/src/datastrategy/cmdqdeviceattrdatastrategy.cpp [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+	if(m_recipe.device == nullptr || m_recipe.data.isEmpty()) {
+		qWarning(CAT_CMDQ_DEVICE_DATA_STRATEGY) << "Invalid arguments, cannot read any data.";
+		return;
+	}
+
+	Command *readDataCommand =
+		new IioDeviceAttributeRead(m_recipe.device, m_recipe.data.toStdString().c_str(), nullptr);
+
+	QObject::connect(readDataCommand, &Command::finished, this, [this](Command *cmd) {
+		IioDeviceAttributeRead *tcmd = dynamic_cast<IioDeviceAttributeRead *>(cmd);
```
Comment (adisuciu) [inline iio-widgets/src/datastrategy/cmdqdeviceattrdatastrategy.cpp:86]: This lambda is pretty long .. should it be mvoed to a separate function ?

File: iio-widgets/src/datastrategy/cmdqdeviceattrdatastrategy.cpp [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+		if(tcmd->getReturnCode() < 0) {
+			qWarning(CAT_CMDQ_DEVICE_DATA_STRATEGY) << "Could not read the value for" << m_recipe.data;
+			return;
+		}
+
+		char *currentValue = tcmd->getResult();
+		m_dataRead = QString(currentValue);
+
+		if(!m_recipe.constDataOptions.isEmpty()) {
```
Comment (adisuciu) [inline iio-widgets/src/datastrategy/cmdqdeviceattrdatastrategy.cpp:99]: Can this logic be simplified ?

File: iio-widgets/src/datastrategy/cmdqchannelattrdatastrategy.cpp [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+	if(m_recipe.channel == nullptr || m_recipe.data.isEmpty()) {
+		qWarning(CAT_CMDQ_CHANNEL_DATA_STATEGY) << "Invalid arguments, cannot read any data.";
+		return;
+	}
+
+	Command *readDataCommand =
+		new IioChannelAttributeRead(m_recipe.channel, m_recipe.data.toStdString().c_str(), nullptr);
+
+	QObject::connect(readDataCommand, &Command::finished, this, [this](Command *cmd) {
+		IioChannelAttributeRead *tcmd = dynamic_cast<IioChannelAttributeRead *>(cmd);
```
Comment (adisuciu) [inline iio-widgets/src/datastrategy/cmdqchannelattrdatastrategy.cpp:85]: This lambda is pretty long. Should it be moved to a function ?

File: iio-widgets/src/datastrategy/cmdqchannelattrdatastrategy.cpp [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+		if(!tcmd) {
+			return;
+		}
+
+		if(tcmd->getReturnCode() < 0) {
```
Comment (adisuciu) [inline iio-widgets/src/datastrategy/cmdqchannelattrdatastrategy.cpp:90]: Can this logic be simplified ?

**Review by adisuciu — CHANGES_REQUESTED** [review]

File: iio-widgets/src/iiowidgetfactory.cpp [hunk 5]
```diff
@@ -195,6 +198,22 @@ IIOWidget *IIOWidgetFactory::buildSingle(uint32_t hint, IIOWidgetFactoryRecipe r
 		dataStrategy = new DeviceAttrDataStrategy(recipe, parent);
 	} else if(hint & ContextAttrData) {
 		dataStrategy = new ContextAttrDataStrategy(recipe, parent);
+	} else if(hint & CMDQAttrData) {
+		auto conn = recipe.connection;
+		if(conn) {
+			dataStrategy = new CmdQChannelAttrDataStrategy(recipe, conn->commandQueue(), parent);
+		} else {
```
Comment (AlexandraTrifan) [inline iio-widgets/src/iiowidgetfactory.cpp:205]: Do you still need to pass the command queue as a constructor param? After the last changes the recipe has a reference to the Connection obj, so it also has a reference to the command queue.

**Review by AlexandraTrifan — APPROVED** [review]

**Review by adisuciu — APPROVED** [review]

**Pull request merged**